buildbot_nix/__init__.py

```diff
@@ -59,85 +83,218 @@ def __init__(
-            error = job.get("error")
-            props = Properties()
-            props.setProperty("virtual_builder_name", name, source)
-            props.setProperty("status_name", f"nix-build .#checks.{attr}", source)
-            props.setProperty("virtual_builder_tags", "", source)
-            props.setProperty("report_status", self.report_status, source)
-
-            drv_path = job.get("drvPath")
-            system = job.get("system")
-            out_path = job.get("outputs", {}).get("out")
-
-            props.setProperty("attr", attr, source)
-            props.setProperty("system", system, source)
-            props.setProperty("drv_path", drv_path, source)
-            props.setProperty("out_path", out_path, source)
-            props.setProperty("default_branch", self.project.default_branch, source)
-            # we use this to identify builds when running a retry
-            props.setProperty("build_uuid", str(uuid.uuid4()), source)
-
-            if error is not None:
-                props.setProperty("error", error, source)
-                triggered_schedulers.append((self.skipped_builds_scheduler, props))
-                continue
-
-            if job.get("isCached"):
-                triggered_schedulers.append((self.skipped_builds_scheduler, props))
-                continue
-
-            build_props.setProperty(f"{attr}-out_path", out_path, source)
-            build_props.setProperty(f"{attr}-drv_path", drv_path, source)
-
-            triggered_schedulers.append((self.builds_scheduler, props))
-        return triggered_schedulers
+        all_deps: dict[str, set[str]] = dict()
+        for drv, info in self.drv_info.items():
+            all_deps[drv] = set(info.get("inputDrvs").keys())
+        def closure_of(key: str, deps: dict[str, set[str]]) -> set[str]:
+            r, size = set([key]), 0
+            while len(r) != size:
+                size = len(r)
+                r.update(*[ deps[k] for k in r ])
+            return r.difference([key])
+        # essentially filters out jobs
+        job_set = set(( drv for drv in ( job.drvPath for job in self.jobs ) if drv ))
+
+        all_deps = { k: closure_of(k, all_deps).intersection(job_set) for k in job_set }
+        builds_to_schedule = list(self.jobs)
+        build_schedule_order = []
+        sorter = graphlib.TopologicalSorter(all_deps)
+        for item in sorter.static_order():
+            i = 0
+            while i < len(builds_to_schedule):
+                if item == builds_to_schedule[i].drvPath:
+                    build_schedule_order.append(builds_to_schedule[i])
+                    del builds_to_schedule[i]
+                else:
+                    i += 1
+
+        done = []
+        scheduled: list[Tuple[NixEvalJobSuccess, dict[int, int], defer.Deferred[list[int]]]] = []
+        all_results = SUCCESS
+        ss_for_trigger = self.prepareSourcestampListForTrigger()
+        while len(build_schedule_order) > 0 or len(scheduled) > 0:
```

Comment:
Is there maybe a way to split this function up into smaller pieces?

Reply:
Indeed, i havent done that yet, i want to get the features working then refactor more. Introduce more datatypes and split it up